pkg/watermark/fetch/edge_fetcher.go

```diff
@@ -67,19 +67,27 @@ func NewEdgeFetcher(ctx context.Context, bucketName string, storeWatcher store.W
 		storeWatcher:     storeWatcher,
 		processorManager: manager,
 		lastProcessedWm:  lastProcessedWm,
-		log:              log,
+		log:              log.With("bucket", bucketName),
 	}
 }
 
-// GetWatermark gets the largest possible watermark from all the active processors of all partitions for the given offset and partition.
-// We calculate the watermark for the given offset and partition, update the lastProcessedWatermark of the given partition to the watermark we just calculate.
-// Then, we compare the lastProcessedWatermark from all partitions and return the minimum as the edge watermark.
-// deletes the processor if it's not active.
-func (e *edgeFetcher) GetWatermark(inputOffset isb.Offset, fromPartitionIdx int32) wmb.Watermark {
+// ComputeWatermark processes the offset on the partition indicated and returns the overall Watermark
+// from all Partitions
+func (e *edgeFetcher) ComputeWatermark(offset isb.Offset, fromPartitionIdx int32) wmb.Watermark {
+	err := e.updateWatermark(offset, fromPartitionIdx)
+	if err != nil {
+		return wmb.InitialWatermark
+	}
+	return e.getWatermark()
```

Comment:
Do we need a read lock to access lastProcessedWatermark in `getWatermark`?

Reply:
oh, good catch. It looks like that was being locked by the caller before but then the code was rearranged